ros/src/computing/planning/common/lib/astar_search/astar_search/src/astar_search.cpp

```diff
@@ -0,0 +1,722 @@
+
+  // Convert each point to index and check if the node is Obstacle
+  for (double x = left; x < right; x += resolution)
+  {
+    for (double y = top; y > bottom; y -= resolution)
+    {
+      // 2D point rotation
+      int index_x = (x * cos_theta - y * sin_theta + base_x) / resolution;
+      int index_y = (x * sin_theta + y * cos_theta + base_y) / resolution;
+
+      if (isOutOfRange(index_x, index_y))
+      {
+        return true;
+      }
+      else if (nodes_[index_y][index_x][0].status == STATUS::OBS)
+      {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool AstarSearch::calcWaveFrontHeuristic(const SimpleNode& sn)
+{
+  // Set start point for wavefront search
+  // This is goal for Astar search
+  nodes_[sn.index_y][sn.index_x][0].hc = 0;
+  WaveFrontNode wf_node(sn.index_x, sn.index_y, 1e-10);
+  std::queue<WaveFrontNode> qu;
+  qu.push(wf_node);
+
+  // State update table for wavefront search
+  // Nodes are expanded for each neighborhood cells (moore neighborhood)
+  double resolution = costmap_.info.resolution;
+  static std::vector<WaveFrontNode> updates = {
+    getWaveFrontNode(0, 1, resolution),
+    getWaveFrontNode(-1, 0, resolution),
+    getWaveFrontNode(1, 0, resolution),
+    getWaveFrontNode(0, -1, resolution),
+    getWaveFrontNode(-1, 1, std::hypot(resolution, resolution)),
+    getWaveFrontNode(1, 1, std::hypot(resolution, resolution)),
+    getWaveFrontNode(-1, -1, std::hypot(resolution, resolution)),
+    getWaveFrontNode(1, -1, std::hypot(resolution, resolution)),
+  };
+
+  // Get start index
+  int start_index_x;
+  int start_index_y;
+  int start_index_theta;
+  poseToIndex(start_pose_local_.pose, &start_index_x, &start_index_y, &start_index_theta);
+
+  // Whether the robot can reach goal
+  bool reachable = false;
+
+  // Start wavefront search
+  while (!qu.empty())
+  {
+    WaveFrontNode ref = qu.front();
+    qu.pop();
+
+    WaveFrontNode next;
+    for (const auto& u : updates)
+    {
+      next.index_x = ref.index_x + u.index_x;
+      next.index_y = ref.index_y + u.index_y;
+
+      // out of range OR already visited OR obstacle node
+      if (isOutOfRange(next.index_x, next.index_y) || nodes_[next.index_y][next.index_x][0].hc > 0 ||
+          nodes_[next.index_y][next.index_x][0].status == STATUS::OBS)
+      {
+        continue;
+      }
+
+      // Take the size of robot into account
+      if (detectCollisionWaveFront(next))
+      {
+        continue;
+      }
+
+      // Check if we can reach from start to goal
+      if (next.index_x == start_index_x && next.index_y == start_index_y)
+      {
+        reachable = true;
+      }
+
+      // Set wavefront heuristic cost
+      next.hc = ref.hc + u.hc;
+      nodes_[next.index_y][next.index_x][0].hc = next.hc;
+
+      qu.push(next);
+    }
+  }
+
+  // End of search
+  return reachable;
+}
+
+// Simple collidion detection for wavefront search
+bool AstarSearch::detectCollisionWaveFront(const WaveFrontNode& ref)
```

Comment:
The "detectCollision" function assumed the shape of the vehicle as a rectangle, why is this function assuming a square in this function?

Reply:
No deep meaning. It's experimental function and not used by default.